Add support for AWS profile with role_arn configuration

daaru00 opened the pull request:
## What has been implemented?

**Allow usage of symbolic link for AWS credentials and other files.**

For additional security, using Linux, is a good practice to link files with sensitive content (like credentials) into [Encrypted Private Directory](https://help.ubuntu.com/community/EncryptedPrivateDirectory). These changes allow to use this configurations without fail `fileExistsSync()` check.

**Add [aws-sdk](https://www.npmjs.com/package/aws-sdk) package as dependency**

This is required in order to [assume a role](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) using AWS STS client.

**Makes `loadInstanceCredentials()` asynchronous**

If I have not misunderstood the `loadInstanceCredentials()` function is always called with await operator even though the function was not declared as async. These change will force this declaration in order to wait AWS STS call to retrieve credentials.

**Add support for AWS Profile with assume role functionality**

Closes issue #506

## Steps to verify

- Add into `~/.aws/credentials` a profile section with role_arn
```
[mycompanyprofile]
aws_access_key_id = xxxxxxxxxxxxxxxxxxxx
aws_secret_access_key = xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

[myprofile]
role_arn = arn:aws:iam::000000000000:role/MyOrganizationAccountAccessRole
source_profile = mycompanyprofile
```
- Set environment variable `AWS_PROFILE` or add it into `.env` file
```
AWS_PROFILE=myprofile
```
- Run `serverless deploy` command

## Todos:

- [ ] Write tests
- [x] Write / update documentation
- [x] Run Prettier

daaru00 commented:
Hi @eahefnawy,

> any chance you could get this to work without depending on the aws-sdk?

I removed the `aws-sdk` dependency and used [aws4](https://www.npmjs.com/package/aws4) to sign request with [Signature Version 4](https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html). This kind of signature is too complicated to replicate, and doing the whole process from scratch doesn't seem like a good idea..

Is this solution good enough for you?

eahefnawy commented on src/cli/commands/utils.js:
```diff
@@ -124,9 +209,9 @@ const loadAwsCredentials = () => {
  */
 
-const loadInstanceCredentials = () => {
+const loadInstanceCredentials = async () => {
```
I realize why you had to make that method async, but that could have serious implications on the codebase where this method might be called. I need to search and review the codebase to make sure there are no calls to this method that we've missed.

daaru00 replied:
absolutely understandable, definitely needs some testing :+1: let me know

eahefnawy reviewed (commented):
@daaru00 thank you for making that change. It must've not been easy 😊 

It looks good, I'm just a bit worried about the async change and I wanna test it a bit.

eahefnawy commented:
@daaru00 I checked the codebase and tested this with multiple scenarios. Couldn't break it. Good job 👍 

If you could just add docs in the [credentials section](https://github.com/serverless/components#credentials), this PR would be G2M!

daaru00 commented:
@eahefnawy that's great! 

I added the documentation with extra section on credentials for AWS, let me know if it's clear enough.

While I was writing some details about session durationI also added the possibility to modify it using a custom environment variable. This in order to to minimize the effect of expired credentials during "watch" command. I guess does not re-execute this portion of code again, am I wrong?

eahefnawy reviewed (approved):
Yeah, I think that's fine.

Thanks again @daaru00 ... G2M 🚀
